cilium/CFP-33716-distributed-packet-tracing.md

```diff
@@ -0,0 +1,586 @@
+- `<type>` corresponds to the IP option value ([https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml](https://www.iana.org/assignments/ip-parameters/ip-parameters.xhtml))
+- By default, the feature is disabled (`<type>=0`).
+- `<type>=0` disables this feature.
+
+```go
+EnableIPOptionTracing = "enable-ip-option-tracing"
+```
+
+```go
+if option.Config.EnableIPOptionTracing > 0 {
+    cDefinesMap["ENABLE_PACKET_IP_TRACING"] = 
+    strconv.Itoa(option.Config.EnableIPOptionTracing)
+}
+```
+
+### Section 3: Parsing
+
+Parsing at the ingress point of the TC hook allows for early detection and tracing of a flow.
+
+```c
+#ifdef ENABLE_PACKET_IP_TRACING
+    check_and_store_ip_trace_id(ctx, ENABLE_PACKET_IP_TRACING);
+    // Do parsing and storing of ip_trace_id
+#endif
+```
+```c
+// Function to parse trace ID and store it if valid
+static __always_inline void check_and_store_ip_trace_id(struct __ctx_buff *ctx, __u8 ip_opt_type_value) {
+    __s64 trace_id;
+    trace_id = trace_id_from_ctx(ctx, ip_opt_type_value); // do parsing
+
+    // check if the trace ID is valid
+    if (trace_id > 0) {
+        bpf_trace_id_set(trace_id);
+    } else {
+        bpf_trace_id_set(0);
+    }
+}
+```
+At ingress, the function `trace_id_from_ctx` is called to parse the IP options from incoming packets. If the packet is valid, it proceeds to parse the IP options to extract the trace ID.
+
+We loop over the IP options to extract the trace ID. This loop is unrolled to a maximum of three iterations to minimize performance impact and meet BPF requirements. The maximum of three iterations accounts for any combination of trace ID, user-provided options, and either egress NAT or DSR options. If we read an IP option from the trace type which we configured (see Section 2), then this is stored as the `ip_trace_id`.
+
+Several edge cases are considered during parsing:
+- **No IP Options:** If the packet header indicates there are no options, parsing stops early, avoiding unnecessary processing.
+- **Invalid `ip_trace_id`:** invalid IDs are handled by setting the trace ID to 0.
+
+### Section 4: `ip_trace_id` storing
+
+Creating a per-CPU array-map allows for storage of each packet’s `ip_trace_id` as it is processed. When the feature gate is disabled (see Section 2: Feature enablement), the map is not rendered, avoiding space inefficiencies.
+
+```c
+#ifdef ENABLE_PACKET_IP_TRACING 
+struct {
+    __uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+    __uint(max_entries, 1);
+    __type(key, __u32); // only one key
+    __type(value, __u64); // ip_trace_id type
+    __uint(pinning, LIBBPF_PIN_BY_NAME);
+} ip_trace_id_map __section_maps_btf;
+...
+// remaining declarations 
+#endif
+```
+
+**Pros**
+- Non-competitive storage; since this is not saved directly to a predefined field in the skb struct, there aren’t competing uses for this storage place. It is dedicated directly to the ip_trace_id of the packet being processed.
```

Comment:
(minor) suggest: "Packet tracing ID is stored outside of the skb struct which has limited room for custom attributes."

Reply:
Fixed.